src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReadyToRunHeaderNode.cs

```diff
@@ -5,33 +5,31 @@
 
 using Internal.Runtime;
 using Internal.Text;
-using Internal.TypeSystem;
+
+using Debug = System.Diagnostics.Debug;
 
 namespace ILCompiler.DependencyAnalysis
 {
     public class ReadyToRunHeaderNode : ObjectNode, ISymbolDefinitionNode
     {
         private struct HeaderItem
         {
-            public HeaderItem(ReadyToRunSectionType id, ObjectNode node, ISymbolNode startSymbol, ISymbolNode endSymbol)
+            public HeaderItem(ReadyToRunSectionType id, ObjectNode node)
             {
                 Id = id;
                 Node = node;
-                StartSymbol = startSymbol;
-                EndSymbol = endSymbol;
             }
 
             public readonly ReadyToRunSectionType Id;
             public readonly ObjectNode Node;
-            public readonly ISymbolNode StartSymbol;
-            public readonly ISymbolNode EndSymbol;
         }
 
         private List<HeaderItem> _items = new List<HeaderItem>();
 
-        public void Add(ReadyToRunSectionType id, ObjectNode node, ISymbolNode startSymbol, ISymbolNode endSymbol = null)
+        public void Add(ReadyToRunSectionType id, ObjectNode node)
         {
-            _items.Add(new HeaderItem(id, node, startSymbol, endSymbol));
+            Debug.Assert(node is ISymbolDefinitionNode);
```

Comment:
Out of curiosity, wouldn't it make sense to ensure this through generic constraint?

```csharp
public void Add<TObjectNode>(ReadyToRunSectionType id, TObjectNode node)
    where TObjectNode : ObjectNode, ISymbolDefinitionNode
```

Reply:
Yup, good point!